### PR TITLE
Support installing sensu plugins and extensions

### PR DESCRIPTION
**Pull request by treydock**
# Pull Request Checklist

## Description

Add both type and provider and class to manage sensu plugin and extension installation.

Had to have apt based systems have packages depend on `Class[apt::update]` as adding a second repo caused `apt-get update` to take place later and prevented install of core sensu-go packages.  The dependency change ensures `apt-get update` is run before trying to install any packages.

## Motivation and Context

sensu-go docs for plugins: https://docs.sensu.io/sensu-go/5.2/installation/plugins/

This change makes it easy to utilize plugins.

## General

- [x] Update `README.md` with any necessary configuration snippets

- [x] New parameters are documented

- [x] New parameters have tests

- [x] Tests pass - `bundle exec rake validate lint spec`

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -70,11 +63,21 @@
 
   case $facts['os']['family'] {
     'RedHat': {
+      $os_package_require = []
```
Comment (ghoneycutt) [inline manifests/init.pp]: does this not work if it is set to `undef` ?

File: manifests/init.pp [hunk 2]
```diff
@@ -70,11 +63,21 @@
...
     default: {
       fail("Detected osfamily <${::osfamily}>. Only RedHat and Debian are supported.")
     }
   }
+
+  if $manage_repo {
+    include ::sensu::repo
+    $package_require = [Class['::sensu::repo']] + $os_package_require
+  } else {
+    $package_require = undef
```
Comment (ghoneycutt) [inline manifests/init.pp]: Perhaps we should document which class(es) this `$package_require` variable is meant to be used in. Since it is just specified here and not used, it makes the code harder to understand.

File: manifests/plugins.pp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#   Default is OS dependent.
+# @param plugins
+#   Plugins to install
+# @param extensions
+#   Extensions to install
+#
+class sensu::plugins (
+  String $package_ensure = 'installed',
+  String $package_name = 'sensu-plugins-ruby',
+  Array $dependencies = [],
```
Comment (ghoneycutt) [inline manifests/plugins.pp]: We should not use empty assignments such as `''`, `[]` and `{}` and instead use `undef`.

File: manifests/plugins.pp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  Variant[Array, Hash] $plugins = {},
+  Variant[Array, Hash] $extensions = {},
+) {
+
+  include ::sensu
+
+  if $::sensu::manage_repo {
```
Comment (ghoneycutt) [inline manifests/plugins.pp]: this seems to duplicate the code already existing in the sensu class.

File: manifests/repo/community.pp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        baseurl         => "https://packagecloud.io/sensu/community/el/${repo_release}/\$basearch",
+        descr           => 'sensu_community',
+        enabled         => 1,
+        gpgcheck        => 0,
+        gpgkey          => 'https://packagecloud.io/sensu/community/gpgkey',
+        metadata_expire => 300,
+        repo_gpgcheck   => 1,
+        sslcacert       => '/etc/pki/tls/certs/ca-bundle.crt',
+        sslverify       => 1,
+}
```
Comment (ghoneycutt) [inline manifests/repo/community.pp]: style, i think this paran is misaligned..

File: manifests/repo/community.pp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        include  => {
+          'src' => true,
+        },
+        key      => {
+          'id'     => '7F54E8A5C0CB51DBE612D2F50156BD72FEC8CD59',
+          'source' => 'https://packagecloud.io/sensu/community/gpgkey',
+        },
+      }
+    }
+    default: {
```
Comment (ghoneycutt) [inline manifests/repo/community.pp]: perhaps the default should be to emit a `warning()`

File: data/os/RedHat.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+sensu::plugins::dependencies:
```
Comment (ghoneycutt) [inline data/os/RedHat.yaml]: the dependencies are missing from the spec tests

File: manifests/init.pp (hunk 1 above)
Comment (treydock) [reply manifests/init.pp]: I do array merge later so wanted all values to be arrays to keep thing simpler.

File: manifests/repo/community.pp (hunk 6 above)
Comment (treydock) [reply manifests/repo/community.pp]: It's not possible to ever hit this default, it's just there for lint.  The `sensu` class will fail if `osfamily` is not `RedHat` or `Debian`.

File: manifests/plugins.pp (hunk 3 above)
Comment (treydock) [reply manifests/plugins.pp]: So using `undef` would require nesting a lot of logic inside an `if` statements which could make the rest of manifest a bit harder to read without changing behavior.  Still want `undef` as default?  Also in case of `$dependencies` the real default is specified in Hiera data.

File: manifests/repo/community.pp (hunk 5 above)
Comment (treydock) [reply manifests/repo/community.pp]: It's not, at least doesn't look like it in VIM.  Everything lines up with `metadata_expire` which is longest parameter name.

File: manifests/init.pp (hunk 2 above)
Comment (treydock) [reply manifests/init.pp]: Add comment about where `$package_require` is used.

File: data/os/RedHat.yaml (hunk 7 above)
Comment (treydock) [reply data/os/RedHat.yaml]: Added these to unit tests.

File: manifests/repo/community.pp (hunk 6 above)
Comment (ghoneycutt) [reply manifests/repo/community.pp]: That's what should happen :)

File: manifests/plugins.pp (hunk 3 above)
Comment (ghoneycutt) [reply manifests/plugins.pp]: Yes, please use `undef` and update the data types. I think it will only introduce one layer of an `if` statement around the current code.

File: spec/classes/plugins_spec.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
+require 'spec_helper'
+
+describe 'sensu::plugins', :type => :class do
+  on_supported_os({facterversion: '3.8.0'}).each do |os, facts|
+    if facts[:osfamily] == 'Debian'
+      package_require = ['Class[Sensu::Repo::Community]', 'Class[Apt::Update]']
```
Comment (ghoneycutt) [inline spec/classes/plugins_spec.rb]: What I've done in the past is outside the describe set the data such as

```ruby
platforms => {
  debian => {
    :package_require => ['Class[Sensu::Repo::Community]', 'Class[Apt::Update]'],
    :dependencies => ['make','gcc','g++','libssl-dev'],
  },
  el => {
    :package_require => ['Class[Sensu::Repo::Community]'],
    :dependencies => ['make','gcc','gcc-c++','openssl-devel'],
  },
}
```

Then keep your if statement and use that to assign the data

```
package_require = platforms[#{facts[:osfamily]}[:package_require]
```

This is useful as the data changes and we add more platforms. With the current code you assume if osfamily is not Debian then it must be RedHat. By using the above method, it makes it very clear which platform's data is being used.

File: manifests/plugins.pp (hunk 3 above)
Comment (treydock) [reply manifests/plugins.pp]: Updated to default to `undef`.

File: manifests/init.pp (hunk 1 above)
Comment (ghoneycutt) [reply manifests/init.pp]: makes sense to me, especially since this is a private variable and not a param

File: manifests/init.pp [hunk 9]
```diff
@@ -70,11 +63,23 @@
 
   case $facts['os']['family'] {
     'RedHat': {
+      $os_package_require = []
     }
     'Debian': {
+      $os_package_require = [Class['::apt::update']]
     }
     default: {
       fail("Detected osfamily <${::osfamily}>. Only RedHat and Debian are supported.")
```
Comment (ghoneycutt) [inline manifests/init.pp]: I dont think osfamily is used anymore in lieu of `$facts['os']['family']`

File: manifests/plugins.pp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+# @param package_ensure
+#   Ensure property for sensu plugins package.
+# @param package_name
+#   Name of the Sensu plugins ruby package.
+# @param dependencies
+#   Package dependencies needed to install plugins and extensions.
+#   Default is OS dependent.
+# @param plugins
+#   Plugins to install
+# @param extensions
```
Comment (ghoneycutt) [inline manifests/plugins.pp]: what's a plugin vs an extension?

File: manifests/repo/community.pp (hunk 5 above)
Comment (ghoneycutt) [reply manifests/repo/community.pp]: check you aren't using hard tabs..  this paran is definitely off as it should line up with `yumrepo`

File: manifests/repo/community.pp (hunk 6 above)
Comment (ghoneycutt) [reply manifests/repo/community.pp]: To avoid the empty case statement, suggest doing two if statements

```puppet
if $facts['os']['family'] == 'RedHat' {
  ...
}

if $facts['os']['family'] == 'Debian' {
  ...
}
```

This is also easier to read than a if-elsif

File: spec/acceptance/04_plugins_spec.rb [hunk 11]
```diff
@@ -0,0 +1,78 @@
+require 'spec_helper_acceptance'
+
+describe 'sensu::plugins class', unless: RSpec.configuration.sensu_cluster do
+  agent = hosts_as('sensu_agent')[0]
+  backend = hosts_as('sensu_backend')[0]
+  context 'on agent' do
+    it 'should work without errors' do
```
Comment (ghoneycutt) [inline spec/acceptance/04_plugins_spec.rb]: The `it` text should be updated to show that we're checking idempotency.

`it should work without errors and be idempotent` or you can wrap that idempotency check like this

```ruby
it 'should be idempotent' do
  apply_manifest_on(agent, pp, :catch_changes  => true)
end
```

File: spec/classes/plugins_spec.rb (hunk 8 above)
Comment (treydock) [reply spec/classes/plugins_spec.rb]: Took similar approach but defined a function in spec_helper that can be used by all the unit tests so these are all in one place.

File: manifests/plugins.pp (hunk 10 above)
Comment (treydock) [reply manifests/plugins.pp]: I'm not sure but the command `sensu-install` makes a distinction as plugins use `--plugin` flag and extensions use `--extension` flag and gem names have different prefixes.

File: manifests/init.pp (hunk 9 above)
Comment (treydock) [reply manifests/init.pp]: Changed to match how fact is referenced for actual case statement to use `$facts['os']['family']`.

File: spec/acceptance/04_plugins_spec.rb (hunk 11 above)
Comment (treydock) [reply spec/acceptance/04_plugins_spec.rb]: Used the `it should work without errors and be idempotent`.

**Comment by treydock** [comment]
Tests are failing because Debian lacks `sensu-plugins-ruby` package for Debian.  See https://github.com/sensu/sensu-go/issues/2757

**Comment by ghoneycutt** [comment]
Thanks for creating the ticket with sensu-go. Suggest skipping those tests with Debian and adding a comment like

```
# TODO: package is missing on Debian - See https://github.com/sensu/sensu-go/issues/2757
```

**Comment by treydock** [comment]
Add commit to skip all plugin related tests for Debian.

**Comment by treydock** [comment]
Added commit that documents `sensu::plugins` in README.

File: README.md [hunk 12]
```diff
@@ -173,6 +174,40 @@ class { 'sensu':
 }
 ```
 
+### Installing Plugins
+
+Plugin installation is handled by the `sensu::plugins` class.
```
Comment (ghoneycutt) [inline README.md]: s/installation/management

Installation is a non-idempotent term

File: README.md [hunk 13]
```diff
@@ -173,6 +174,40 @@ class { 'sensu':
 }
 ```
 
+### Installing Plugins
```
Comment (ghoneycutt) [inline README.md]: Need a similar section for Managing Extensions

File: README.md (hunk 12 above)
Comment (treydock) [reply README.md]: Change made

File: spec/acceptance/04_plugins_spec.rb [hunk 14]
```diff
@@ -0,0 +1,83 @@
+require 'spec_helper_acceptance'
+
+describe 'sensu::plugins class', unless: RSpec.configuration.sensu_cluster do
+  agent = hosts_as('sensu_agent')[0]
+  backend = hosts_as('sensu_backend')[0]
+  before do
+    if fact_on(agent, 'operatingsystem') == 'Debian'
+      skip("TODO: package is missing on Debian - See https://github.com/sensu/sensu-go/issues/2757")
```
Comment (ghoneycutt) [inline spec/acceptance/04_plugins_spec.rb]: Should probably just link to https://github.com/sensu/sensu-plugins-omnibus/issues/3

File: spec/acceptance/sensu_plugin_spec.rb [hunk 15]
```diff
@@ -0,0 +1,78 @@
+require 'spec_helper_acceptance'
+
+describe 'sensu_plugin', if: RSpec.configuration.sensu_full do
+  agent = hosts_as('sensu_agent')[0]
+  before do
+    if fact_on(agent, 'operatingsystem') == 'Debian'
+      skip("TODO: package is missing on Debian - See https://github.com/sensu/sensu-go/issues/2757")
```
Comment (ghoneycutt) [inline spec/acceptance/sensu_plugin_spec.rb]: Should probably just link to https://github.com/sensu/sensu-plugins-omnibus/issues/3

File: README.md (hunk 13 above)
Comment (treydock) [reply README.md]: Added section.

File: spec/acceptance/04_plugins_spec.rb (hunk 14 above)
Comment (treydock) [reply spec/acceptance/04_plugins_spec.rb]: Updated

File: spec/acceptance/sensu_plugin_spec.rb (hunk 15 above)
Comment (treydock) [reply spec/acceptance/sensu_plugin_spec.rb]: Updated

File: manifests/repo/community.pp (hunk 6 above)
Comment (treydock) [reply manifests/repo/community.pp]: Implemented.  I'll open separate pull request to do this in `sensu::repo`.

File: manifests/repo/community.pp (hunk 5 above)
Comment (treydock) [reply manifests/repo/community.pp]: I fixed that when switching to if statements.

File: lib/puppet/provider/sensu_plugin/sensu_install.rb [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+  commands :sensu_install => 'sensu-install'
+
+  def self.instances
+    plugins = []
+
+    output = gem('list', '--local', '^sensu-(plugins|extensions)')
+    Puppet.debug("gem output: #{output}")
+    output.each_line do |o|
+      next unless o.start_with?('sensu-')
+      if o =~ /^(\S+)\s+\((.+)\)/
```
Comment (ghoneycutt) [inline lib/puppet/provider/sensu_plugin/sensu_install.rb]: please document this regex in plain english

File: README.md [hunk 17]
```diff
@@ -173,6 +175,70 @@ class { 'sensu':
...
```
Comment (ghoneycutt) [inline README.md]: If you can specify plugins and extensions as absent, this should be `### Managing Plugins`

File: README.md [hunk 18]
```diff
@@ -173,6 +175,70 @@ class { 'sensu':
...
+
+Plugin management is handled by the `sensu::plugins` class.
+
+Example installing plugins on agent:
+
+```puppet
```
Comment (ghoneycutt) [inline README.md]: should show how to disable the dependencies as many folks will want to do this because the packages can conflict

File: lib/puppet/provider/sensu_plugin/sensu_install.rb [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+      end
+    end
+  end
+
+  def self.latest_versions
+    return @latest_versions if @latest_versions
+    @latest_versions = {}
+    output = gem('search', '--remote', '--all', "^sensu-(plugins|extensions)-")
+    output.each_line do |o|
+      if o =~ /^(\S+)\s+\((.+)\)/
```
Comment (ghoneycutt) [inline lib/puppet/provider/sensu_plugin/sensu_install.rb]: document regex

File: lib/puppet/type/sensu_plugin.rb [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+@summary Manages Sensu plugins
+@example Install a sensu plugin
+  sensu_plugin { 'disk-checks':
+    ensure  => 'present',
+  }
+
+@example Install latest version of a sensu plugin
+  sensu_plugin { 'disk-checks':
+    ensure  => 'present',
+    version => 'latest',
```
Comment (ghoneycutt) [inline lib/puppet/type/sensu_plugin.rb]: example should show how to pin to a specific version as sensu recommends that you pin your check versions

File: manifests/plugins.pp [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+#   Extensions to install
+#
+class sensu::plugins (
+  String $package_ensure = 'installed',
+  String $package_name = 'sensu-plugins-ruby',
+  Optional[Array] $dependencies = undef,
+  Optional[Variant[Array, Hash]] $plugins = undef,
+  Optional[Variant[Array, Hash]] $extensions = undef,
```
Comment (ghoneycutt) [inline manifests/plugins.pp]: Seems that VP is advocating using `[]` and `{}`. Should this be changed?

File: lib/puppet/provider/sensu_plugin/sensu_install.rb (hunk 16 above)
Comment (treydock) [reply lib/puppet/provider/sensu_plugin/sensu_install.rb]: Added comments.

File: README.md (hunk 17 above)
Comment (treydock) [reply README.md]: Added docs about setting `absent` for both plugins and extensions.

File: README.md (hunk 18 above)
Comment (treydock) [reply README.md]: Add docs to plugins section about passing `dependencies => []`.

File: lib/puppet/provider/sensu_plugin/sensu_install.rb (hunk 19 above)
Comment (treydock) [reply lib/puppet/provider/sensu_plugin/sensu_install.rb]: Added comments

File: lib/puppet/type/sensu_plugin.rb (hunk 20 above)
Comment (treydock) [reply lib/puppet/type/sensu_plugin.rb]: Added example of passing specific version.

File: manifests/plugins.pp (hunk 21 above)
Comment (treydock) [reply manifests/plugins.pp]: I switched all 3 from using `Optional` to defaulting to empty Array.
